Prefer Bash arrays to POSIX splitting when possible

We use this pattern in some shell scripts for building up a list of arguments:

```sh
ARGS=
...
ARGS="$ARGS nextarg"
...
# shellcheck disable=SC2086 # (This word splitting is intentional.)
somecommand $ARGS
```

When the scripts are Bash-specific (`#!/usr/bin/env bash`), it's cleaner and more robust to use real arrays:

```sh
ARGS=()
...
ARGS+=("nextarg")
...
somecommand "${ARGS[@]}"
```

`git grep -P '\b(\w+)="\$\1 '` turns up at least some.

--- a/contrib/coverage.bash
+++ b/contrib/coverage.bash
@@ -28,12 +28,12 @@ while [[ $# -gt 0 ]]; do
 			;;
 		--jobs)
 			shift
-			make_args+=("-j" "$1")
-			runtests_args+=("--jobs" "$1")
+			make_args+=(-j "$1")
+			runtests_args+=(--jobs "$1")
 			;;
 		--os)
 			shift
-			runtests_args+=("--os" "$1")
+			runtests_args+=(--os "$1")
 			;;
 		*)
 			echo "$(basename "$0"): unknown option '$1'"

--- a/test/asm/test.sh
+++ b/test/asm/test.sh
@@ -61,9 +61,9 @@ else
 fi
 
 for i in *.asm notexist.asm; do
-	RGBASMFLAGS="-Weverything -Bcollapse"
+	RGBASMFLAGS=(-Weverything -Bcollapse)
 	if [ -f "${i%.asm}.flags" ]; then
-		RGBASMFLAGS="$RGBASMFLAGS @${i%.asm}.flags"
+		RGBASMFLAGS+=("@${i%.asm}.flags")
 	fi
 	for variant in '' ' piped'; do
 		(( tests++ ))
@@ -79,7 +79,7 @@ for i in *.asm notexist.asm; do
 			desired_errname=/dev/null
 		fi
 		if [ -z "$variant" ]; then
-			"$RGBASM" $RGBASMFLAGS -o "$o" "$i" >"$output" 2>"$errput"
+			"$RGBASM" "${RGBASMFLAGS[@]}" -o "$o" "$i" >"$output" 2>"$errput"
 			desired_output=$desired_outname
 			desired_errput=$desired_errname
 		else
@@ -96,7 +96,7 @@ for i in *.asm notexist.asm; do
 			# that's harder to deal with and was broken when the feature was
 			# first implemented.
 			# shellcheck disable=SC2002
-			cat "$i" | "$RGBASM" $RGBASMFLAGS -o "$o" - >"$output" 2>"$errput"
+			cat "$i" | "$RGBASM" "${RGBASMFLAGS[@]}" -o "$o" - >"$output" 2>"$errput"
 
 			# Use two otherwise unused files for temp storage
 			desired_output="$input"
@@ -162,10 +162,10 @@ done
 # These tests do their own thing
 
 i="invalid-source-date-epoch"
-RGBASMFLAGS="-Weverything -Bcollapse"
+RGBASMFLAGS=(-Weverything -Bcollapse)
 (( tests++ ))
 echo "${bold}${green}${i}...${rescolors}${resbold}"
-SOURCE_DATE_EPOCH=0x1234 "$RGBASM" $RGBASMFLAGS /dev/null >"$output" 2>"$errput"
+SOURCE_DATE_EPOCH=0x1234 "$RGBASM" "${RGBASMFLAGS[@]}" /dev/null >"$output" 2>"$errput"
 tryDiff /dev/null "$output" out
 our_rc=$?
 tryDiff invalid-source-date-epoch.err "$errput" err
@@ -177,15 +177,15 @@ fi
 
 evaluateDepTest () {
 	i="$1"
-	RGBASMFLAGS="-Weverything -Bcollapse -M -"
+	RGBASMFLAGS=(-Weverything -Bcollapse -M -)
 	if [ -f "$i/a.flags" ]; then
-		RGBASMFLAGS="$RGBASMFLAGS @$i/a.flags"
+		RGBASMFLAGS+=("@$i/a.flags")
 	fi
 	# Piping the .asm file to rgbasm would not make sense for dependency generation,
 	# so just test the normal variant
 	(( tests++ ))
 	echo "${bold}${green}${i}...${rescolors}${resbold}"
-	"$RGBASM" $RGBASMFLAGS -o "$o" "$i"/a.asm >"$output" 2>"$errput"
+	"$RGBASM" "${RGBASMFLAGS[@]}" -o "$o" "$i"/a.asm >"$output" 2>"$errput"
 
 	fixed_output="$input"
 	if type -t cygpath >/dev/null; then
@@ -223,15 +223,15 @@ else
 	state_outname="$o"
 fi
 state_features="  all  " # Test trimming whitespace
-RGBASMFLAGS="-Weverything -Bcollapse"
+RGBASMFLAGS=(-Weverything -Bcollapse)
 for variant in '' '.pipe'; do
 	(( tests++ ))
 	echo "${bold}${green}${i%.asm}${variant}...${rescolors}${resbold}"
 	if [ -z "$variant" ]; then
-		"$RGBASM" $RGBASMFLAGS -s "$state_features:$state_outname" "$i"/a.asm >"$output" 2>"$errput"
+		"$RGBASM" "${RGBASMFLAGS[@]}" -s "$state_features:$state_outname" "$i"/a.asm >"$output" 2>"$errput"
 	else
 		# shellcheck disable=SC2002
-		cat "$i"/a.asm | "$RGBASM" $RGBASMFLAGS -s "$state_features:$state_outname" - >"$output" 2>"$errput"
+		cat "$i"/a.asm | "$RGBASM" "${RGBASMFLAGS[@]}" -s "$state_features:$state_outname" - >"$output" 2>"$errput"
 	fi
 
 	tryDiff /dev/null "$output" out

--- a/test/asm/test.sh
+++ b/test/asm/test.sh
@@ -11,8 +11,7 @@ input="$(mktemp)"
 output="$(mktemp)"
 errput="$(mktemp)"
 
-# Immediate expansion is the desired behavior.
-# shellcheck disable=SC2064
+# shellcheck disable=SC2064 # (Immediate expansion is the desired behavior.)
 trap "rm -f ${o@Q} ${gb@Q} ${input@Q} ${output@Q} ${errput@Q}" EXIT
 
 tests=0
@@ -91,11 +90,7 @@ for i in *.asm notexist.asm; do
 				continue
 			fi
 
-			# Stop! This is not a Useless Use Of Cat. Using cat instead of
-			# stdin redirection makes the input an unseekable pipe - a scenario
-			# that's harder to deal with and was broken when the feature was
-			# first implemented.
-			# shellcheck disable=SC2002
+			# shellcheck disable=SC2002 # (This use of `cat` intentionally makes the input an unseekable pipe.)
 			cat "$i" | "$RGBASM" "${RGBASMFLAGS[@]}" -o "$o" - >"$output" 2>"$errput"
 
 			# Use two otherwise unused files for temp storage
@@ -230,7 +225,7 @@ for variant in '' '.pipe'; do
 	if [ -z "$variant" ]; then
 		"$RGBASM" "${RGBASMFLAGS[@]}" -s "$state_features:$state_outname" "$i"/a.asm >"$output" 2>"$errput"
 	else
-		# shellcheck disable=SC2002
+		# shellcheck disable=SC2002 # (This use of `cat` intentionally makes the input an unseekable pipe.)
 		cat "$i"/a.asm | "$RGBASM" "${RGBASMFLAGS[@]}" -s "$state_features:$state_outname" - >"$output" 2>"$errput"
 	fi
 

--- a/test/fix/test.sh
+++ b/test/fix/test.sh
@@ -3,8 +3,7 @@
 export LC_ALL=C
 
 tmpdir="$(mktemp -d)"
-# Immediate expansion is the desired behavior.
-# shellcheck disable=SC2064
+# shellcheck disable=SC2064 # (Immediate expansion is the desired behavior.)
 trap "cd; rm -rf ${tmpdir@Q}" EXIT
 
 src="$PWD"
@@ -73,10 +72,7 @@ runTest () {
 			eval "$RGBFIX" $flags out.gb '>out.out' '2>out.err'
 			subst=out.gb
 		elif [[ "$variant" = ' piped' ]]; then
-			# Stop! This is not a Useless Use Of Cat. Using cat instead of
-			# stdin redirection makes the input an unseekable pipe - a scenario
-			# that's harder to deal with.
-			# shellcheck disable=SC2002
+			# shellcheck disable=SC2002 # (This use of `cat` intentionally makes the input an unseekable pipe.)
 			cat "$desired_input" | eval "$RGBFIX" $flags - '>out.gb' '2>out.err'
 			subst='<stdin>'
 		elif [[ "$variant" = ' output' ]]; then

--- a/test/gfx/test.sh
+++ b/test/gfx/test.sh
@@ -5,8 +5,7 @@
 
 errtmp="$(mktemp)"
 
-# Immediate expansion is the desired behavior.
-# shellcheck disable=SC2064
+# shellcheck disable=SC2064 # (Immediate expansion is the desired behavior.)
 trap "rm -f ${errtmp@Q} result.{png,1bpp,2bpp,pal,tilemap,attrmap,palmap} out*.png" EXIT
 
 tests=0

--- a/test/link/test.sh
+++ b/test/link/test.sh
@@ -10,8 +10,7 @@ outtemp="$(mktemp)"
 outtemp2="$(mktemp)"
 outtemp3="$(mktemp)"
 
-# Immediate expansion is the desired behavior.
-# shellcheck disable=SC2064
+# shellcheck disable=SC2064 # (Immediate expansion is the desired behavior.)
 trap "rm -f ${otemp@Q} ${gbtemp@Q} ${gbtemp2@Q} ${outtemp@Q} ${outtemp2@Q} ${outtemp3@Q}" EXIT
 
 tests=0

--- a/test/link/test.sh
+++ b/test/link/test.sh
@@ -93,9 +93,9 @@ for i in *.asm; do
 	startTest
 	"$RGBASM" -o "$otemp" "${test}.asm"
 
-	RGBLINKFLAGS=""
+	RGBLINKFLAGS=()
 	if [ -f "${test}.flags" ]; then
-		RGBLINKFLAGS="$RGBLINKFLAGS @${test}.flags"
+		RGBLINKFLAGS+=("@${test}.flags")
 	fi
 
 	# Some tests have variants depending on flags
@@ -103,14 +103,14 @@ for i in *.asm; do
 	for flag in '-d' '-t' '-w'; do
 		if [ -f "${test}-no${flag}.out" ]; then
 			continueTest "-no${flag}"
-			rgblinkQuiet $RGBLINKFLAGS -o "$gbtemp" "$otemp" 2>"$outtemp"
+			rgblinkQuiet "${RGBLINKFLAGS[@]}" -o "$gbtemp" "$otemp" 2>"$outtemp"
 			tryDiff "${test}-no${flag}.out" "$outtemp"
 			evaluateTest
 			ran_flag=true
 		fi
 		if [ -f "${test}${flag}.out" ]; then
 			continueTest "$flag"
-			rgblinkQuiet $RGBLINKFLAGS ${flag} -o "$gbtemp" "$otemp" 2>"$outtemp"
+			rgblinkQuiet "${RGBLINKFLAGS[@]}" ${flag} -o "$gbtemp" "$otemp" 2>"$outtemp"
 			tryDiff "${test}${flag}.out" "$outtemp"
 			evaluateTest
 			ran_flag=true
@@ -125,7 +125,7 @@ for i in *.asm; do
 		[[ -e "$script" ]] || break # If the glob doesn't match, it just... doesn't expand!
 
 		continueTest "${script#${test}}"
-		rgblinkQuiet $RGBLINKFLAGS -l "$script" -o "$gbtemp" "$otemp" 2>"$outtemp"
+		rgblinkQuiet "${RGBLINKFLAGS[@]}" -l "$script" -o "$gbtemp" "$otemp" 2>"$outtemp"
 		tryDiff "${script%.link}.out" "$outtemp"
 		evaluateTest
 		ran_flag=true
@@ -136,7 +136,7 @@ for i in *.asm; do
 
 	# The rest of the tests just links a file, and maybe checks the binary
 	continueTest
-	rgblinkQuiet $RGBLINKFLAGS -o "$gbtemp" "$otemp" 2>"$outtemp"
+	rgblinkQuiet "${RGBLINKFLAGS[@]}" -o "$gbtemp" "$otemp" 2>"$outtemp"
 	tryDiff "${test}.out" "$outtemp"
 	bin=${test}.out.bin
 	if [ -f "$bin" ]; then

